rust/tw_proto/src/ffi.rs

```diff
@@ -0,0 +1,104 @@
+#![allow(clippy::missing_safety_doc)]
+
+use std::borrow::Cow;
+use std::ffi::c_char;
+use tw_memory::c_string_standalone;
+use tw_memory::ffi::c_byte_array::{CByteArray, CByteArrayResult};
+use tw_memory::ffi::c_byte_array_ref::CByteArrayRef;
+use tw_memory::ffi::c_result::ErrorCode;
+
+#[repr(C)]
+pub enum CProtoError {
+    Ok = 0,
+    Internal = 1,
+    ErrorDeserializingMsg = 2,
+    ErrorSerializingMsg = 3,
+}
+
+impl From<CProtoError> for ErrorCode {
+    fn from(e: CProtoError) -> Self {
+        e as ErrorCode
+    }
+}
+
+/// Takes a serialized `Ethereum::Proto::SigningInput` message, deserializes it and returns a serialized message back.
+/// This FFI is used within integration tests.
+/// \param input *non-null* byte array.
+/// \param input_len length of the input byte array.
+/// \return C-compatible result with a C-compatible byte array.
+#[no_mangle]
+pub unsafe extern "C" fn pass_eth_signing_msg_through(
+    input: *const u8,
+    input_len: usize,
+) -> CByteArrayResult {
+    let data = CByteArrayRef::new(input, input_len)
+        .to_vec()
+        .unwrap_or_default();
+
+    let msg: crate::Ethereum::Proto::SigningInput = match crate::deserialize(&data) {
+        Ok(msg) => msg,
+        Err(_) => return CByteArrayResult::error(CProtoError::ErrorDeserializingMsg),
+    };
+    crate::serialize(&msg)
+        .map(CByteArray::from)
+        .map_err(|_| CProtoError::ErrorSerializingMsg)
+        .into()
+}
+
+/// Returns an encoded and signed `polkadot_test_signing_input` transaction.
+/// This FFI is used within integration tests.
+/// \return C-compatible, null-terminated string.
+#[no_mangle]
+pub unsafe extern "C" fn polkadot_tx_expected_encoded() -> *const c_char {
+    c_string_standalone("29028488dc3417d5058ec4b4503e0c12ea1a0a89be200fe98922423d4334014fa6b0ee003d91a06263956d8ce3ce5c55455baefff299d9cb2bb3f76866b6828ee4083770b6c03b05d7b6eb510ac78d047002c1fe5c6ee4b37c9c5a8b09ea07677f12e50d3200000005008eaf04151687736326c9fea17e25fc5287613693c912909cb226aa4794f26a48e5c0")
```

Comment:
Just curious on how you creates this test data? Also, I would probably not expose stuff like that publicly, respectively move it to `test/`.

Reply:
I've copied this from `PolkadotSigner_SignTransferDOT` C++ example to complete the following check point 🙂 
> Try to build a tx with protobuf in rust, serialize it, get it in C++, sign it in C++ and ensure to have correct result
https://github.com/trustwallet/wallet-core/blob/d67078daa580b37063c97be66a625aaee9664882/tests/chains/Polkadot/SignerTests.cpp#L100